app/post_metadata.go

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package app
+
+import (
+	"strings"
+
+	"github.com/dyatlov/go-opengraph/opengraph"
+	"github.com/mattermost/mattermost-server/model"
+)
+
+func (a *App) PreparePostListForClient(originalList *model.PostList) (*model.PostList, *model.AppError) {
+	list := &model.PostList{
+		Posts: make(map[string]*model.Post),
+		Order: originalList.Order,
+	}
+
+	for id, originalPost := range originalList.Posts {
+		post, err := a.PreparePostForClient(originalPost)
+		if err != nil {
+			return originalList, err
+		}
+
+		list.Posts[id] = post
+	}
+
+	return list, nil
+}
+
+func (a *App) PreparePostForClient(originalPost *model.Post) (*model.Post, *model.AppError) {
```

Comment:
@sudheerDev If you're at all curious, the main logic for this is here. If any of these fields aren't set yet, it goes out and pulls in the data for them. The rest of the changes to the app package are to replace the calls that previously just added proxied image URLs to the post with calls to this new method and to add a new method for getting multiple custom emojis

Reply:
The rest is just bits and pieces of logic to make that easier like logic to count reactions or look for possible emojis in a string